tests/core/data_layer/test_data_rpc.py

```diff
@@ -2638,3 +2639,113 @@ async def test_dl_proof_changed_root(offer_setup: OfferSetup, seeded_random: ran
+        self_hostname, one_wallet_and_one_simulator_services
+    )
+    # Number of farmed blocks to check our batch update was not published.
+    NUM_BLOCKS_WITHOUT_PUBLISH = 10
+    async with init_data_layer(wallet_rpc_port=wallet_rpc_port, bt=bt, db_path=tmp_path) as data_layer:
+        data_rpc_api = DataLayerRpcApi(data_layer)
+        res = await data_rpc_api.create_data_store({})
+        assert res is not None
+
+        store_id = bytes32(hexstr_to_bytes(res["id"]))
+        await farm_block_check_singleton(data_layer, full_node_api, ph, store_id, wallet=wallet_rpc_api.service)
+
+        to_insert = [(b"a", b"\x00\x01"), (b"b", b"\x00\x02"), (b"c", b"\x00\x03")]
+        for key, value in to_insert:
+            changelist: List[Dict[str, str]] = [{"action": "insert", "key": key.hex(), "value": value.hex()}]
+
+            res = await data_rpc_api.batch_update(
+                {"id": store_id.hex(), "changelist": changelist, "publish_on_chain": False}
+            )
+            assert res == {}
+
+            for _ in range(NUM_BLOCKS_WITHOUT_PUBLISH):
+                await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
+            keys_values = await data_rpc_api.get_keys_values({"id": store_id.hex()})
+            assert keys_values == {"keys_values": []}
+            pending_root = await data_layer.data_store.get_pending_root(tree_id=store_id)
+            assert pending_root is not None
+            assert pending_root.status == Status.PENDING_BATCH
+
+        key = b"d"
+        value = b"\x00\x04"
+        to_insert.append((key, value))
+
+        changelist = [{"action": "insert", "key": key.hex(), "value": value.hex()}]
+        res = await data_rpc_api.batch_update({"id": store_id.hex(), "changelist": changelist})
+        update_tx_rec0 = res["tx_id"]
+        await farm_block_with_spend(full_node_api, ph, update_tx_rec0, wallet_rpc_api)
+
+        keys_values = await data_rpc_api.get_keys_values({"id": store_id.hex()})
+        assert len(keys_values["keys_values"]) == len(to_insert)
+        kv_dict = {item["key"]: item["value"] for item in keys_values["keys_values"]}
+        for key, value in to_insert:
+            assert kv_dict["0x" + key.hex()] == "0x" + value.hex()
+        prev_keys_values = keys_values
+
+        key = b"e"
+        value = b"\x00\x05"
+        changelist = [{"action": "insert", "key": key.hex(), "value": value.hex()}]
+        res = await data_rpc_api.batch_update(
+            {"id": store_id.hex(), "changelist": changelist, "publish_on_chain": False}
+        )
+        assert res == {}
+
+        for _ in range(NUM_BLOCKS_WITHOUT_PUBLISH):
+            await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
+
+        await data_rpc_api.clear_pending_roots({"store_id": store_id.hex()})
+        pending_root = await data_layer.data_store.get_pending_root(tree_id=store_id)
+        assert pending_root is None
```

Comment:
Also check the latest confirmed root.  I believe there is a bug here where it will be the second to last unpublished batch.